framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,531 @@
+     *
+     * @param basename the simple name of the class (without the package), in Outer$Inner form
+     * @param classname the fully-qualified name of the class in question, so that it can be printed
+     *     as an enum if it is one
+     * @param aClass the AClass representing the classname
+     * @return the number of outer classes within which this class is nested
+     */
+    private int printClassDefinitions(String basename, String classname, AClass aClass) {
+
+        String nameToPrint = basename;
+        String rest = "";
+        if (basename.contains(".")) {
+            nameToPrint = basename.substring(0, basename.indexOf('.'));
+            rest = basename.substring(basename.indexOf('.') + 1);
+        }
+
+        // Enums cannot be nested within non-classes,
+        // so only print as an enum if the leaf has been reached.
+        if (enumConstants.containsKey(classname) && "".equals(rest)) {
+            printWriter.print("enum");
+        } else {
+            printWriter.print("class ");
+        }
+        printAnnotations(aClass.tlAnnotationsHere);
+        printWriter.print(" " + nameToPrint);
+        printTypeParameters(classname);
+        printWriter.println(" {");
+        if ("".equals(rest)) {
+            return 0;
+        } else {
+            return 1 + printClassDefinitions(rest, classname, aClass);
+        }
+    }
+
+    /**
+     * The implementation of {@link #write(AScene, Map, Map, Map, Writer)} and {@link #write(AScene,
+     * Map, Map, Map, String)}. Prints imports, classes, method signatures, and fields in stub file
+     * format, all with appropriate annotations.
+     *
+     * @param scene the scene to write
+     */
+    private void writeImpl(AScene scene) {
+        // Write out all imports
+        ImportDefCollector importDefCollector;
+        try {
+            importDefCollector = new ImportDefCollector(scene);
+        } catch (DefException e) {
+            throw new BugInCF(e.getMessage(), e);
+        }
+        importDefCollector.visit();
+        printWriter.println();
+
+        // For each class
+        for (Map.Entry<String, AClass> classEntry : scene.classes.entrySet()) {
+            String classname = classEntry.getKey();
+            AClass aClass = classEntry.getValue();
+            String pkg = packagePart(classname);
+            if (!"".equals(pkg)) {
+                printWriter.println("package " + pkg + ";");
+            }
+            String basename = basenamePart(classname);
+            if ("package-info".equals(basename) || "module-info".equals(basename)) {
+                continue;
+            }
+
+            // At this point, we no longer care about the distinction between packages
+            // and inner classes, so we should replace the $ in the definition of any
+            // inner classes with a ., so that they are printed correctly in stub files.
+            classname = classname.replace('$', '.');
```

Comment:
It's confusing for a single variable to contain different types of data on different lines.  I suggest using a new variable (with a different name) to distinguish the binary name from the fully-qualified name (or whatever the two formats are).